### PR TITLE
feat: ActionTrace export contract (#94) + property-based invariant tests (#99)

### LINKED ISSUE #94: Interop: export action traces for downstream lesson extraction
## Context

agent-kernel already records action traces and has roadmap work around OpenTelemetry export. A downstream learning layer can only be useful if traces preserve enough information about actions, policies, denials, corrections, and outcomes.

## Goal

Document or implement a stable export shape for action traces that can be consumed by external analysis tools, including LessonWeaver-style lesson extraction.

## Scope

1. Review the existing `ActionTrace` / trace store model.
2. Define the minimal export fields needed for downstream analysis:
   - trace id
   - action/tool name
   - input/output summary or redacted payload reference
   - policy decision
   - error or denial reason
   - timestamps
   - sensitivity tags
   - optional human correction metadata if available
3. Add a JSON export example.
4. Add a short doc explaining how this differs from full observability export.
5. Keep privacy and policy metadata explicit.

## Acceptance criteria

- There is a documented trace export shape.
- The shape is stable enough for downstream adapters.
- Sensitive payloads are not exported by default.
- The example includes at least one allowed action and one denied or failed action.
- The design aligns with existing OpenTelemetry work instead of competing with it.

## Notes for AI coding agents

Check issue #38 before implementing. Prefer documenting the export contract first if the runtime API is not yet ready.

### LINKED ISSUE #99: Add property-based tests for policy bypass and indirect capability use
## Context

In agent/tool systems, the dangerous failures are not always direct unauthorized calls. They can be indirect paths where a capability becomes reachable through context, handles, scoped tokens, chained tools, or policy gaps.

agent-kernel already owns authorization, policy enforcement, firewalling and audit responsibilities. This makes it the right place to test policy invariants under adversarial or unusual execution conditions.

## Problem

Traditional unit tests cover expected policy paths, but may miss edge cases such as:

- token scope confusion
- handle expansion outside original grant constraints
- denied capabilities still producing usable frames
- policy traces leaking raw sensitive argument values
- indirect tool use bypassing intended authorization
- response modes exposing more than intended

## Proposal

Add property-based tests for policy/capability invariants.

These tests should generate varied principals, capability requests, scopes, constraints, response modes, handles and policy decisions, then assert that core security/audit properties always hold.

## Candidate properties

- No tool invocation occurs without a policy decision.
- No denied capability produces a usable frame.
- No handle expansion exceeds the original grant constraints.
- No capability token works outside its scope.
- No expired/single-use token can be reused.
- No policy trace contains raw sensitive argument values.
- No raw tool output reaches the LLM-safe frame path unless explicitly allowed.
- Every allowed/denied decision has a stable audit explanation.

## Acceptance criteria

- [ ] Add property-based tests for at least three high-value policy invariants.
- [ ] Include generated variations of principals, roles, scopes and constraints.
- [ ] Include at least one indirect-use or handle-expansion scenario.
- [ ] Ensure failures are reproducible with seed output.
- [ ] Document which invariants are being tested and why.
- [ ] Avoid brittle assertions on human-readable messages when stable codes exist.

## Non-goals

- Full formal verification.
- Replacing existing unit tests.
- Testing every possible third-party policy engine.

## Related ecosystem work

- dgenio/weaver-spec#72

### PR DESCRIPTION
**Pull request by dgenio**
Implements the recommended combined-PR group from issue triage: **#94** (export action traces) and **#99** (property-based tests). The two reinforce each other — #94 defines a stable, redaction-safe trace contract and #99 proves the audit/policy invariants that contract relies on.

### #94 — Action trace export contract
- **`export_action_trace` / `export_action_traces`** (`trace.py`): a stable, versioned (`TRACE_EXPORT_VERSION = "1"`), JSON-serialisable shape for `ActionTrace` records so downstream tools (e.g. LessonWeaver-style lesson extraction) can consume the audit trail without depending on internals.
- Derived **only** from already-redaction-safe trace fields — `args` (memory payloads stripped at record time) and `result_summary` (post-firewall counts/flags) — so exporting cannot widen the I-01 boundary. A *denied* request never produces a trace (policy gates before invoke, I-02), so `status` is `succeeded`/`failed`; denials surface via `explain_denial`.
- `ActionTrace` now records the invoked capability's **`sensitivity`** (plumbed through the invoke + stream paths).
- Optional human-correction metadata can be attached at export time.
- New [`docs/trace_export.md`](docs/trace_export.md) (incl. how it differs from the OpenTelemetry export) + [`examples/trace_export_demo.py`](examples/trace_export_demo.py) (one succeeded + one failed action), wired into `make ci`.

### #99 — Property-based invariant tests
- New `tests/test_policy_properties.py` (Hypothesis) asserting, across generated principals/capabilities/scopes/constraints/handles/tokens:
  - every allow/deny carries a **stable reason code**;
  - `max_rows` **never exceeds the policy cap**;
  - **handle expansion never exceeds the original grant** (indirect-use scenario);
  - tokens **never verify outside their scope**; tampered/expired tokens are always rejected;
  - **policy traces never leak raw scope values**;
  - the **trace export is always JSON-serialisable**.
- Adds `hypothesis` as a dev dependency.

## Why
Closes the audit/trace gap in #94 (the `TraceStore` had no export contract) and the test-rigor gap in #99 (no property-based coverage). Grouped because they share the audit/trace surface (`trace.py`, `ActionTrace`, redaction) and the export's redaction guarantee is exactly what the property tests verify.

## How verified
- `ruff check src/ tests/ examples/` — clean
- `ruff format --check` — 77 files already formatted
- `mypy src/` — Success: no issues found in 41 source files
- `pytest` — **581 passed, 1 skipped** (`test_mcp_driver` skipped: `mcp` not installable in this sandbox due to an unrelated PyJWT/RECORD conflict; unaffected by this change)
- Example list (incl. `trace_export_demo.py`) runs clean

## Risks / caveats
- `ActionTrace` gains a `sensitivity` field (defaults to `NONE`); traces constructed directly keep working. No retro-compat concern per the request.
- Adds `hypothesis` to the dev extras only (no runtime dependency added).

Scope: limited to #94 + #99. Closes #94, closes #99.

https://claude.ai/code/session_01A1SNbC3izrsX4JggWWVSAB

---
_Generated by [Claude Code](https://claude.ai/code/session_01A1SNbC3izrsX4JggWWVSAB)_

### CODE REVIEW

File: src/weaver_kernel/models.py [hunk 1]
```diff
@@ -413,6 +413,14 @@ class ActionTrace:
     args: dict[str, Any]
     response_mode: ResponseMode
     driver_id: str
+    sensitivity: SensitivityTag = SensitivityTag.NONE
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/models.py]: Good catch — this is addressed in `b2be0e0`. `sensitivity` is now declared **last** on `ActionTrace` (after `result_summary`), so it no longer shifts the positional `__init__` order of the pre-existing public fields; it still defaults to `SensitivityTag.NONE`. The field docstring notes this rationale explicitly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/trace_export.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  ]
+}
+```
+
+## Stability
+
+`TRACE_EXPORT_VERSION` is bumped only on a **breaking** change to the field
+shape. New optional fields may be added without a bump, so consumers should
+ignore unknown keys. Assert on `status`, `sensitivity`, and `reason`/`error`
+rather than on human-readable strings, which may evolve.
```
[the remark this answers is not in the document]
Comment (dgenio) [reply docs/trace_export.md]: Fixed in `b2be0e0`. The Stability section no longer references a `reason` field — it now reads "Assert on `status`, `sensitivity`, and the presence of `error`", matching the documented export schema (which only carries `error`).

---
_Generated by [Claude Code](https://claude.ai/code)_
